payment/src/main/java/org/killbill/billing/payment/core/janitor/IncompletePaymentAttemptTask.java

```diff
@@ -155,18 +256,135 @@ public boolean apply(final PaymentTransactionModelDao input) {
+                                                        @Nullable final Integer attemptNumber,
+                                                        @Nullable final UUID userToken,
+                                                        final InternalTenantContext internalTenantContext) throws LockFailedException {
+        // First, fix the transaction itself
+        final TransactionStatus latestTransactionStatus = incompletePaymentTransactionTask.updatePaymentAndTransactionIfNeeded(accountId,
+                                                                                                                               paymentTransactionId,
+                                                                                                                               currentTransactionStatus,
+                                                                                                                               paymentTransactionInfoPlugin,
+                                                                                                                               attemptNumber,
+                                                                                                                               userToken,
+                                                                                                                               internalTenantContext);
+        final boolean hasTransactionChanged = latestTransactionStatus == null;
+
+        // Don't insert a notification for the on-the-fly Janitor
+        final boolean shouldInsertNotification = attemptNumber != null;
+        if (!hasTransactionChanged && shouldInsertNotification) {
+            insertNewNotificationForUnresolvedTransactionIfNeeded(paymentTransactionId,
+                                                                  latestTransactionStatus,
+                                                                  attemptNumber,
+                                                                  userToken,
+                                                                  internalTenantContext.getAccountRecordId(),
+                                                                  internalTenantContext.getTenantRecordId());
+        }
+
+        // If there is a payment attempt associated with that transaction, we need to update it as well
+        boolean hasAttemptChanged = false;
+        final PaymentTransactionModelDao paymentTransactionModelDao = paymentDao.getPaymentTransaction(paymentTransactionId, internalTenantContext);
+        if (paymentTransactionModelDao.getAttemptId() != null) {
+            final PaymentAttemptModelDao paymentAttemptModelDao = paymentDao.getPaymentAttempt(paymentTransactionModelDao.getAttemptId(), internalTenantContext);
+            if (paymentAttemptModelDao != null) {
+                if (hasTransactionChanged || retrySMHelper.getInitialState().getName().equals(paymentAttemptModelDao.getStateName())) {
```

Comment:
What about the `RETRIED` state?

Reply:
I don't think it matters (`RETRIED` is a terminal state): either the transaction changed, and we need to re-run the completion code, or the attempt was stuck in `INIT`.

I have a subsequent PR which will verify the behavior for `ABORTED` though (non-terminal state).